perf/src/stats.rs

```diff
@@ -0,0 +1,368 @@
+use hdrhistogram::Histogram;
+use quinn::StreamId;
+use std::sync::atomic::{AtomicBool, AtomicU64, AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::time::{Duration, Instant, SystemTime};
+
+pub struct Stats {
+    /// Test start time
+    start_instant: Instant,
+    /// Test start system time
+    start: SystemTime,
+    /// Durations of uploads
+    upload_duration: Histogram<u64>,
+    /// Durations of downloads
+    download_duration: Histogram<u64>,
+    /// Time from finishing the upload until receiving the first byte of the response
+    fbl: Histogram<u64>,
+    /// Throughput for uploads
+    upload_throughput: Histogram<u64>,
+    /// Throughput for downloads
+    download_throughput: Histogram<u64>,
+    /// The total amount of requests executed
+    requests: usize,
+    /// Stats accumulated over each interval
+    intervals: Vec<Interval>,
+}
+
+impl Default for Stats {
+    fn default() -> Self {
+        Self {
+            start_instant: Instant::now(),
+            start: SystemTime::now(),
+            upload_duration: Histogram::new(3).unwrap(),
+            download_duration: Histogram::new(3).unwrap(),
+            fbl: Histogram::new(3).unwrap(),
+            upload_throughput: Histogram::new(3).unwrap(),
+            download_throughput: Histogram::new(3).unwrap(),
+            requests: 0,
+            intervals: vec![],
+        }
+    }
+}
+
+impl Stats {
+    pub fn on_interval(&mut self, start: Instant, stream_stats: &mut Vec<Arc<StreamStats>>) {
+        let mut interval = Interval::new(start - self.start_instant, self.start_instant.elapsed());
+
+        stream_stats.retain(|stream_stats| {
+            self.record(stream_stats.clone());
+            interval.record_stream_stats(stream_stats.clone());
+            // Retain if not finished yet
+            stream_stats.finished.load(Ordering::Relaxed) == false
```

Comment:
While not a huge deal here since the runtimes are single threaded, I usually go with `Ordering::SeqCst` until I can take some time and reason about the correct semantics. In this case I would do the same to reduce the risk of weird bugs.